### PR TITLE
Add test behavior to yapf_code.sh

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
Please pay particular attention to the use of `set -f` - is there a better way to do what I'm accomplishing with that?

### CODE REVIEW

File: tools/distrib/yapf_code.sh [hunk 1]
```diff
@@ -34,28 +34,40 @@ set -ex
...
-$PYTHON -m pip install futures
+$PYTHON -m pip install --upgrade pip
+$PYTHON -m pip install --upgrade futures
 $PYTHON -m pip install yapf==0.16.0
 
-exclusion_args=""
-for exclusion in $EXCLUSIONS; do
-  exclusion_args="$exclusion_args --exclude $exclusion"
-done
+function yapf {
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: delete the `function` keyword and write `yapf()` instead of `yapf`

File: tools/distrib/yapf_code.sh [hunk 2]
```diff
@@ -34,28 +34,40 @@ set -ex
 cd $(dirname $0)/../..
 
 DIRS=src/python
-EXCLUSIONS='src/python/grpcio/grpc_*.py src/python/grpcio_health_checking/grpc_*.py src/python/grpcio_reflection/grpc_*.py src/python/grpcio_tests/grpc_*.py'
+EXCLUSIONS='grpcio/grpc_*.py grpcio_health_checking/grpc_*.py grpcio_reflection/grpc_*.py grpcio_tests/grpc_*.py'
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: this should be an array like:

```sh
EXCLUSIONS=(
    'grpcio/grpc_*.py'
    'grpcio_health_checking/grpc_*.py'
    ....
)
```

then see below for updated usage

File: tools/distrib/yapf_code.sh [hunk 3]
```diff
@@ -34,28 +34,40 @@ set -ex
...
+$PYTHON -m pip install --upgrade pip
+$PYTHON -m pip install --upgrade futures
 $PYTHON -m pip install yapf==0.16.0
 
-exclusion_args=""
-for exclusion in $EXCLUSIONS; do
-  exclusion_args="$exclusion_args --exclude $exclusion"
-done
+function yapf {
+    for exclusion in $EXCLUSIONS; do
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: this should do:

```sh
local exclusion exclusion_args=()
for exclusion in "${EXCLUSIONS[@]}"; do
  exclusion_args+=( "--exclude" "$1/${exclusion}" )
done
$PYTHON -m yapf -i -r --style=setup.cfg -p "${exclusion_args[@]}" $1
```

then you can delete the `set -f` below

File: tools/distrib/yapf_code.sh [hunk 4]
```diff
@@ -34,28 +34,40 @@ set -ex
...
-  tempdir=`mktemp -d`
-  cp -RT $dir $tempdir
-  $PYTHON -m yapf -i -r -p $exclusion_args $dir
-  if ! diff -r $dir $tempdir; then
-    script_result=1
-  fi
-  rm -rf $tempdir
-done
-exit $script_result
+if [ "$TEST" == "" ]
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: do instead:

`if [ -z "${TEST}" ]; then`

File: tools/distrib/yapf_code.sh [hunk 5]
```diff
@@ -34,28 +34,40 @@ set -ex
...
+then
+    for dir in $DIRS; do
+	yapf $dir
+    done
+else
+    ok=yes
+    for dir in $DIRS; do
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: `DIRS` should also be an array like the exclusions suggestions above, then all of these can be properly quoted

File: tools/distrib/yapf_code.sh [hunk 6]
```diff
@@ -34,28 +34,40 @@ set -ex
...
+	tempdir=`mktemp -d`
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: don't use backticks.  write this: `tempdir=$(mktemp -d)`.

File: tools/distrib/yapf_code.sh [hunk 7]
```diff
@@ -34,28 +34,40 @@ set -ex
...
+	cp -RT $dir $tempdir
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: all of these path vars are missing quoting which can lead to really bad behavior if `$tempdir` contains a space

File: tools/distrib/yapf_code.sh [hunk 8]
```diff
@@ -34,28 +34,40 @@ set -ex
...
+	yapf $tempdir
+	diff -ru $dir $tempdir || ok=no
+	rm -rf $tempdir
+    done
+    if [ $ok == no ]
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: if you're going to use `[`, then you should use `=` and not `==`.  then again, if you're in bash, you should use `[[` everywhere and never use `[`.

File: tools/distrib/yapf_code.sh (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 3 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 4 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 5 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 6 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 7 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 8 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

**Comment by nathanielmanistaatgoogle** [comment]
Thank you very much for your comments @vapier! The file now passes `shellcheck` and the one thing that's surprising to me in the current draft is the nested double quotes in `cd "$(dirname "${0}")/../.."` - those aren't a problem?

**Comment by vapier** [comment]
quotes nest sanely inside of subshells.  so you can do:

`var="output: $(ls "this is one arg"; echo "OK!")"`

File: tools/distrib/yapf_code.sh [hunk 9]
```diff
@@ -31,31 +31,49 @@
...
+else
+    ok=yes
+    for dir in "${DIRS[@]}"; do
+	tempdir=$(mktemp -d)
+	cp -RT "${dir}" "${tempdir}"
+	yapf "${tempdir}"
+	diff -ru "${dir}" "${tempdir}" || ok=no
+	rm -rf "${tempdir}"
+    done
+    if [[ $ok == no ]]
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: use braces with ${ok}

also, existing code suggests we should cuddle up the `then` to this line

File: tools/distrib/yapf_code.sh [hunk 10]
```diff
@@ -31,31 +31,49 @@
...
-  tempdir=`mktemp -d`
-  cp -RT $dir $tempdir
-  $PYTHON -m yapf -i -r -p $exclusion_args $dir
-  if ! diff -r $dir $tempdir; then
-    script_result=1
-  fi
-  rm -rf $tempdir
-done
-exit $script_result
+if [ -z "${TEST}" ]; then
```
Comment (vapier) [inline tools/distrib/yapf_code.sh]: use `[[` here instead of `[` for consistency

File: tools/distrib/yapf_code.sh (hunk 9 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

File: tools/distrib/yapf_code.sh (hunk 10 above)
Comment (nathanielmanistaatgoogle) [inline tools/distrib/yapf_code.sh]: Done.

**Comment by nathanielmanistaatgoogle** [comment]
Thanks again @vapier and back over to you.

**Comment by vapier** [comment]
i could post some more style nits, but i don't think they deviate from existing style in this codebase, so i wouldn't really be complaining about your code/this CL

also, i'm not related to this project at all, so i can't approve it :).

**Comment by nathanielmanistaatgoogle** [comment]
Pre-merge test assessment:

* Basic Tests Linux: [9831](https://github.com/grpc/grpc/issues/9831), [10092](https://github.com/grpc/grpc/issues/10092).
* Basic Tests Mac: [9804](https://github.com/grpc/grpc/issues/9804).
* Bazel Full Build & Tests: Not yet in effect.
* gRPC_interop_pull_requests: [10056](https://github.com/grpc/grpc/issues/10056).

**Pull request merged**